chain/epoch-manager/src/genesis.rs

```diff
@@ -0,0 +1,178 @@
+use std::collections::HashMap;
+use std::iter;
+use std::sync::Arc;
+
+use itertools::Itertools;
+use near_primitives::chains::{MAINNET, TESTNET};
+use near_primitives::epoch_block_info::BlockInfo;
+use near_primitives::epoch_info::{EpochInfo, EpochInfoV2, RngSeed};
+use near_primitives::epoch_manager::EpochConfig;
+use near_primitives::errors::EpochError;
+use near_primitives::hash::CryptoHash;
+use near_primitives::types::validator_stake::ValidatorStake;
+use near_primitives::types::{AccountId, Balance, EpochId, NumSeats, ValidatorId};
+use near_primitives::version::PROD_GENESIS_PROTOCOL_VERSION;
+use rand::{RngCore, SeedableRng};
+use rand_hc::Hc128Rng;
+
+use crate::EpochManager;
+use crate::proposals::find_threshold;
+use crate::validator_selection::proposals_to_epoch_info;
+
+impl EpochManager {
+    pub fn initialize_genesis_epoch_info(
+        &mut self,
+        validators: Vec<ValidatorStake>,
+    ) -> Result<(), EpochError> {
+        let genesis_epoch_info = self.make_genesis_epoch_info(validators)?;
+        // Dummy block info.
+        // Artificial block we add to simplify implementation: dummy block is the
+        // parent of genesis block that points to itself.
+        // If we view it as block in epoch -1 and height -1, it naturally extends the
+        // EpochId formula using T-2 for T=1, and height field is unused.
+        let block_info = Arc::new(BlockInfo::default());
+        let mut store_update = self.store.store_update();
+        self.save_epoch_info(&mut store_update, &EpochId::default(), Arc::new(genesis_epoch_info))?;
+        self.save_block_info(&mut store_update, block_info)?;
+        store_update.commit()?;
+        Ok(())
+    }
+
+    fn make_genesis_epoch_info(
+        &self,
+        validators: Vec<ValidatorStake>,
+    ) -> Result<EpochInfo, EpochError> {
+        // Missing genesis epoch, means that there is no validator initialize yet.
+        let genesis_protocol_version = self.reward_calculator.genesis_protocol_version;
+        let genesis_epoch_config = self.config.for_protocol_version(genesis_protocol_version);
+        let validator_reward =
+            HashMap::from([(self.reward_calculator.protocol_treasury_account.clone(), 0u128)]);
+
+        // use custom code for genesis protocol version 29 used in mainnet and testnet
+        if genesis_protocol_version == PROD_GENESIS_PROTOCOL_VERSION {
+            let genesis_epoch_info =
+                Self::prod_genesis(&genesis_epoch_config, [0; 32], validators, validator_reward);
```

Comment:
To ensure correctness, I checked that the output from `Self::prod_genesis` matched the output from `proposals_to_epoch_info`. Once the output matched, I added in the hash checks for digest so that it doesn't change again in the future.